langchain/agents/agent_toolkits/openapi/spec.py

```diff
@@ -0,0 +1,114 @@
+"""Experimental (not tested) code to simplify an openapi spec for retrieving into a LM's context.
+"""
+
+from dataclasses import dataclass
+from typing import List, Tuple, Union
+
+
+def dereference_refs(spec_obj: dict, full_spec: dict):
+    """Try to substitute $refs. The goal is to get the complete docs for each endpoint in context for now.
+
+    In the few OpenAPI specs I studied, $refs referenced models (or in OpenAPI terms, components)
+    and could be nested. This code most likely misses lots of cases.
+    """
+    def retrieve_ref_path(path: str, full_spec: dict):
+        components = path.split('/')
+        if components[0] != '#':
+            raise RuntimeError("All $refs I've seen so far are uri fragments (start with hash).")
+        out = full_spec
+        for component in components[1:]:
+            out = out[component]
+        return out
+
+    def _dereference_refs(obj: Union[dict, list]) -> Union[dict, list]:
```

Comment:
Stylistic: What do you think about pulling out the inner function definition to the global namespace?

It's not easy to see immediately on which argument it relies on from the outer scope (e.g., `full_spec`)

Reply:
this inner functions just for recursing, makes little sense in any context but this function scope so thought to have here